homeassistant/components/ambient_station/__init__.py

```diff
@@ -0,0 +1,214 @@
+
+    await hass.config_entries.async_forward_entry_unload(
+        config_entry, 'sensor')
+
+    return True
+
+
+class AmbientStation:
+    """Define a class to handle the Ambient websocket."""
+
+    def __init__(
+            self, hass, config_entry, client, monitored_conditions,
+            unit_system):
+        """Initialize."""
+        self._client = client
+        self._config_entry = config_entry
+        self._hass = hass
+        self._lock = RLock()
+        self._ws_reconnect_handle = None
+        self.monitored_conditions = monitored_conditions
+        self.stations = {}
+        self.unit_system = unit_system
+
+        hass.bus.async_listen_once(
+            EVENT_HOMEASSISTANT_STOP, client.websocket.disconnect())
+
+    async def ws_connect(self):
+        """Register handlers and connect to the websocket."""
+        from aioambient.errors import WebsocketError
+
+        def on_connect():
+            """Define a handler to fire when the websocket is connected."""
+            _LOGGER.info('Connected to websocket')
+
+        def on_data(data):
+            """Define a handler to fire when the data is received."""
+            with self._lock:
+                _LOGGER.debug('New data received: %s', data)
+                self.stations[data['macAddress']][ATTR_LAST_DATA] = data
+                async_dispatcher_send(self._hass, TOPIC_UPDATE)
+
+        def on_disconnect():
+            """Define a handler to fire when the websocket is disconnected."""
+            _LOGGER.info('Disconnected from websocket')
+
+        def on_subscribed(data):
+            """Define a handler to fire when the subscription is set."""
+            for station in data['devices']:
+                if station['macAddress'] in self.stations:
+                    continue
+
+                _LOGGER.debug('New station subscription: %s', data)
+
+                self.stations[station['macAddress']] = {
+                    ATTR_LAST_DATA: station['lastData'],
+                    ATTR_LOCATION: station['info']['location'],
+                    ATTR_NAME: station['info']['name'],
+                }
+
+                async_dispatcher_send(self._hass, TOPIC_UPDATE)
+
+                self._hass.async_create_task(
+                    self._hass.config_entries.async_forward_entry_setup(
+                        self._config_entry, 'sensor'))
+
+        self._client.websocket.on_connect(on_connect)
+        self._client.websocket.on_data(on_data)
+        self._client.websocket.on_disconnect(on_disconnect)
+        self._client.websocket.on_subscribed(on_subscribed)
+
+        try:
+            await self._client.websocket.connect()
+        except WebsocketError as err:
+            if self._ws_reconnect_handle is None:
+                _LOGGER.error("Error with the websocket connection: %s", err)
+                self._ws_reconnect_handle = async_track_time_interval(
+                    self._hass, self.ws_connect, timedelta(minutes=2))
```

Comment:
Do we want a back off time increase between attempts?

Reply:
I was attempting to leave a buffer in case something catastrophic had happened; do you think it should be smaller (e.g., 30 seconds)?

Reply:
I was thinking we could start a bit shorter and increase to longer, eg between 30 seconds and 8 min.
```py
delay = min(2*delay, 480)
```